packages/deployments/contracts/contracts/messaging/RootManager.sol

```diff
@@ -33,26 +34,146 @@ contract RootManager is ProposedOwnable, IRootManager, WatcherClient, DomainInde
+
+  /**
+   * @notice Emitted when a proposer is removed
+   * @param proposer The address of the proposer
+   */
+  event ProposerRemoved(address indexed proposer);
+
   event ConnectorAdded(uint32 domain, address connector, uint32[] domains, address[] connectors);
 
   event ConnectorRemoved(uint32 domain, address connector, uint32[] domains, address[] connectors, address caller);
 
   event PropagateFailed(uint32 domain, address connector);
 
+  /**
+   * @notice Emitted when slow mode is activated
+   * @param watcher The address of the watcher who called the function
+   */
+  event SlowModeActivated(address indexed watcher);
+
+  /**
+   * @notice Emitted when optimistic mode is activated
+   */
+  event OptimisticModeActivated();
+
+  /**
+   * @notice Emitted when an optimistic root is propagated
+   * @param aggregateRoot The aggregate root propagated
+   * @param domainsHash The current domain hash
+   */
+  event OptimisticRootPropagated(bytes32 indexed aggregateRoot, bytes32 domainsHash);
+
+  /**
+   * @notice Emitted when a new aggregate root is proposed
+   * @param snapshotId The snapshot id
+   * @param endOfDispute The block at which the dispute period is over
+   * @param aggregateRoot The new aggregate root proposed
+   * @param baseRoot The root of the tree before the snapshot roots were inserted by proposer
+   * @param snapshotsRoots The list of roots added to aggregate tree
+   * @param domains The list of all domains
+   */
+  event AggregateRootProposed(
+    uint256 indexed snapshotId,
+    uint256 endOfDispute,
+    bytes32 indexed aggregateRoot,
+    bytes32 indexed baseRoot,
+    bytes32[] snapshotsRoots,
+    uint32[] domains
+  );
+
+  /**
+   * @notice Emitted when the current proposed root is finalized
+   * @param aggregateRoot The aggregate root finalized
+   */
+  event ProposedRootFinalized(bytes32 aggregateRoot);
+
+  // ============ Errors ============
+
+  error RootManager_proposeAggregateRoot__InvalidSnapshotId(uint256 snapshotId);
+
+  error RootManager_checkDomains__InvalidDomains();
+
+  error RootManager_finalize__InvalidAggregateRoot();
+
+  error RootManager_onlyOptimisticMode__SlowModeOn();
+
+  error RootManager_activateOptimisticMode__OptimisticModeOn();
+
+  error RootManager_aggregate__OptimisticModeOn();
+
+  error RootManager_proposeAggregateRoot__ProposeInProgress();
+
+  error RootManager_finalize__ProposeInProgress();
+
+  error RootManager_onlyProposer__NotWhitelistedProposer(address caller);
+
+  error RootManager_optimisticPropagate__ForbiddenOptimisticRoot();
+
+  error RootManager_slowPropagate__OldAggregateRoot();
+
+  error RootManager_finalize__InvalidInputHash();
+
   // ============ Properties ============
 
+  /**
+   * @notice The number of blocks watchers have to detect and invalidate the proposed root.
+   * @dev    150 blocks is equivalent to 30 minutes in the current Ethereum Proof of Stake.
+   */
+  uint256 public constant DISPUTE_BLOCKS = 150;
```

Comment:
since we will likely want additional hub chains in the future, this should be set via the constructor

Reply:
done.  We also added a setter and introduced a new variables for minDisputeBlock to guard fat-finger problems